src/tribol/interface/tribol.cpp

```diff
@@ -583,83 +725,130 @@ int getCSRMatrix( int** I, int** J, real** vals, int csId,
+   }
+   SLIC_ERROR_ROOT_IF(
+      !coupling_scheme->hasMfemData(), 
+      "No MFEM data exists. The coupling scheme "
+      "must be registered using registerParMesh() to use this method."
+   );
+   return coupling_scheme->getMfemMatrixData()->GetMfemBlockJacobian(
+      *coupling_scheme->getMethodData()
+   );
+}
+
+//------------------------------------------------------------------------------
+void registerMortarGaps( integer meshId,
+                         real * nodal_gaps )
 {
    MeshManager & meshManager = MeshManager::getInstance();
 
-   SLIC_ERROR_IF(!meshManager.hasMesh(meshId), "tribol::registerRealNodalField(): " << 
-                 "no mesh with id, " << meshId << "exists.");
+   SLIC_ERROR_IF(!meshManager.hasMesh(meshId), "tribol::registerMortarGaps(): " << 
+                 "no mesh with id " << meshId << " exists.");
 
    MeshData & mesh = meshManager.GetMeshInstance( meshId );
 
-   switch (field)
+   if (nodal_gaps == nullptr)
    {
-      case MORTAR_GAPS:
-      {
-         if (fieldVariable==nullptr)
-         {
-            SLIC_WARNING( "tribol::registerRealNodalField(): null pointer to data for " << 
-                          "'MORTAR_GAPS' on mesh, " << meshId << ".");
-            mesh.m_isValid = false;
-         }
-         else
-         {
-            mesh.m_nodalFields.m_node_gap = fieldVariable;
-            mesh.m_nodalFields.m_is_node_gap_set = true;
-         }
-         break;
-      }
-      case MORTAR_PRESSURES:
-      {
-         if (fieldVariable==nullptr)
-         {
-            SLIC_WARNING( "tribol::registerRealNodalField(): null pointer to data for " << 
-                          "'MORTAR_PRESSURES' on mesh, " << meshId << ".");
-            mesh.m_isValid = false;
-         }
-         else
-         {
-            mesh.m_nodalFields.m_node_pressure = fieldVariable;
-            mesh.m_nodalFields.m_is_node_pressure_set = true;
-         }
-         break;
-      }
-      default:
-      {
-         SLIC_ERROR( "tribol::registerRealNodalField(): the field argument " << 
-                     "for mesh, " << meshId << ", is not an accepted tribol nodal field." );
-      }
-   } // end switch over field
+      SLIC_WARNING( "tribol::registerMortarGaps(): null pointer to data " << 
+                    "on mesh " << meshId << ".");
+      mesh.m_isValid = false;
+   }
+   else
+   {
+      mesh.m_nodalFields.m_node_gap = nodal_gaps;
+      mesh.m_nodalFields.m_is_node_gap_set = true;
+   }
+   
+}
 
-} // end registerRealNodalField()
+//------------------------------------------------------------------------------
+mfem::ParGridFunction getMfemGap( integer cs_id )
+{
+   auto coupling_scheme = CouplingSchemeManager::getInstance().getCoupling(cs_id);
+   SLIC_ERROR_ROOT_IF(
+      !coupling_scheme->hasMfemSubmeshData(), 
+      "Coupling scheme does not contain MFEM dual field data. "
+      "Was the coupling scheme created with registerParMesh() and is the "
```

Comment:
Update 'registerParMesh()' as required. Reword to tell the use what to do to get this to work instead of posing a question. Elaborate on 'dual field data' so the user has more clarity if this message is tripped. Not everyone will know what the dual field data is referring to and if/when that relates to gaps.

Reply:
Fixed